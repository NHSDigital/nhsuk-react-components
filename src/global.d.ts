declare module 'nhsuk-frontend/packages/components/header/header.js';
declare module 'nhsuk-frontend/packages/components/checkboxes/checkboxes.js';
declare module 'nhsuk-frontend/packages/components/character-count/character-count.js';
declare module 'nhsuk-frontend/packages/components/tabs/tabs.js';
