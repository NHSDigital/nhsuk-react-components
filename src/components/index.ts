export * from './content-presentation/index.js';
export * from './form-elements/index.js';
export * from './layout/index.js';
export * from './navigation/index.js';
export * from './typography/index.js';
export * from './utils/index.js';
