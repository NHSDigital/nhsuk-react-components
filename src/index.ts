export * from './components';
export * from './patterns';
