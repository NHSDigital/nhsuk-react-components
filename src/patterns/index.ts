export * from './nav-a-z/index.js';
export * from './review-date/index.js';
