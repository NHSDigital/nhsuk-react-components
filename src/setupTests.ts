import { TextDecoder, TextEncoder } from 'util';
import '@testing-library/jest-dom';
import { outdent } from 'outdent';

/**
 * Polyfill TextEncoder/TextDecoder for ReactDOM
 *
 * @see {@link https://github.com/jsdom/jsdom/issues/1695}
 */
Object.defineProperties(globalThis, {
  TextEncoder: {
    value: TextEncoder,
    configurable: true,
    enumerable: false,
    writable: true,
  },
  TextDecoder: {
    value: TextDecoder,
    configurable: true,
    enumerable: false,
    writable: true,
  },
});

/**
 * Polyfill Element methods for NHS.UK frontend error summary
 *
 * @see {@link https://github.com/jsdom/jsdom/issues/1695}
 */
Object.defineProperties(Element.prototype, {
  scrollIntoView: { value: jest.fn() },
});

/**
 * Polyfill `matchMedia()` for NHS.UK frontend tabs
 */
Object.defineProperty(globalThis, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: true,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    addEventListener: jest.fn(),
    removeListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

beforeEach(() => {
  const stylesheet = document.createElement('style');

  stylesheet.innerHTML = outdent`
    :root {
      --nhsuk-breakpoint-mobile: 20rem;
      --nhsuk-breakpoint-tablet: 40.0625rem;
      --nhsuk-breakpoint-desktop: 48.0625rem;
      --nhsuk-breakpoint-large-desktop: 61.875rem;
    }
  `;

  // Add styles for NHS.UK frontend checks
  document.head.appendChild(stylesheet);

  // Flag NHS.UK frontend as supported
  document.body.classList.add('nhsuk-frontend-supported');
});
