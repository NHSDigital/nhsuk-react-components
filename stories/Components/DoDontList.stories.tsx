import React from 'react';
import { DoDontList } from '../../src';
import { Meta, StoryObj } from '@storybook/react';

/**
 * This component can be found in the `nhsuk-frontend` repository <a href="https://github.com/nhsuk/nhsuk-frontend/tree/master/packages/components/do-dont-list" target="_blank" rel="noopener noreferrer">here</a>.
 *
 * ## Implementation Notes
 *
 * The `DoDontList` component has one subcomponent: `DoDontList.Item`.
 *
 * As long as a `listType` is supplied to the `DoDontList` component, all subcomponents will render as desired. If you require a `DoDontList.Item` to be different, a `listItemType` prop can be supplied to force the type.
 *
 *
 * The `DoDontList.Item` component can also accept a `prefixText` prop, which can be used to override the default prefix text.
 *
 * See the <b><a href="#custom-prefix-text" >custom prefix text</a></b> story for an example.
 *
 * ## Usage
 *
 * ### Standard
 *
 * ```jsx
 * import { DoDontList } from "nhsuk-react-components";
 *
 * const Element = () => {
 *     return (
 *         <DoDontList listType="do">
 *             <DoDontList.Item>
 *                 cover blisters that are likely to burst with a soft plaster or dressing
 *             </DoDontList.Item>
 *             <DoDontList.Item>wash your hands before touching a burst blister</DoDontList.Item>
 *             <DoDontList.Item>
 *                 allow the fluid in a burst blister to drain before covering it with a plaster or dressing
 *             </DoDontList.Item>
 *         </DoDontList>
 *     );
 * }
 * ```
 */
const meta: Meta<typeof DoDontList> = {
  title: 'Components/DoDontList',
  component: DoDontList,
};
export default meta;
type Story = StoryObj<typeof DoDontList>;

export const Do: Story = {
  render: () => (
    <DoDontList listType="do">
      <DoDontList.Item>
        cover blisters that are likely to burst with a soft plaster or dressing
      </DoDontList.Item>
      <DoDontList.Item>wash your hands before touching a burst blister</DoDontList.Item>
      <DoDontList.Item>
        allow the fluid in a burst blister to drain before covering it with a plaster or dressing
      </DoDontList.Item>
    </DoDontList>
  ),
};

export const Dont: Story = {
  render: () => (
    <DoDontList listType="dont">
      <DoDontList.Item>burst a blister yourself</DoDontList.Item>
      <DoDontList.Item>peel the skin off a burst blister</DoDontList.Item>
      <DoDontList.Item>pick at the edges of the remaining skin</DoDontList.Item>
      <DoDontList.Item>
        wear the shoes or use the equipment that caused your blister until it heals
      </DoDontList.Item>
    </DoDontList>
  ),
};

/**
 * By default DoDontList's of type "dont" will have Items prefixed with `do not`
 *
 * This is the recommended usage, <b><a href="https://service-manual.nhs.uk/design-system/components/do-and-dont-lists#research">as stated in the Service Manual</a></b>
 *
 * However, if you need to override this, you can supply a `prefixText` prop to the `DoDontList.Item` component.
 *
 * This is optional and of type `ReactNode`, so you can supply a string, a JSX element, or `undefined` or `null`
 *
 * See the table below and click the `Show code` button on the story for examples of each.
 *
 * | Value     | Outcome                                     |
 * |-----------|---------------------------------------------|
 * | undefined | The default `do not` text will be displayed |
 * | null      | There will be no prefix                     |
 * | string    | The string will be displayed                |
 * | JSX       | The JSX will be rendered, such as `<span>`  |
 */
export const CustomPrefixText: Story = {
  render: () => (
    <DoDontList listType="dont">
      <DoDontList.Item prefixText="You must not">burst a blister yourself</DoDontList.Item>
      <DoDontList.Item prefixText={undefined}>peel the skin off a burst blister</DoDontList.Item>
      <DoDontList.Item prefixText={null}>pick at the edges of the remaining skin</DoDontList.Item>
      <DoDontList.Item prefixText={<span>please dont </span>}>
        wear the shoes or use the equipment that caused your blister until it heals
      </DoDontList.Item>
      <DoDontList.Item>
        wear the shoes or use the equipment that caused your blister until it heals
      </DoDontList.Item>
    </DoDontList>

  ),
}
