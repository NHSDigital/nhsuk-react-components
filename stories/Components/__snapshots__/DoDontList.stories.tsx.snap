// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/DoDontList CustomPrefixText smoke-test 1`] = `
<div class="nhsuk-do-dont-list">
  <h3 class="nhsuk-do-dont-list__label">
    Don't
  </h3>
  <ul class="nhsuk-list nhsuk-list--cross">
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      You must notburst a blister yourself
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      do not peel the skin off a burst blister
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      pick at the edges of the remaining skin
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      <span>
        please dont
      </span>
      wear the shoes or use the equipment that caused your blister until it heals
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      do not wear the shoes or use the equipment that caused your blister until it heals
    </li>
  </ul>
</div>
`;

exports[`Components/DoDontList Do smoke-test 1`] = `
<div class="nhsuk-do-dont-list">
  <h3 class="nhsuk-do-dont-list__label">
    Do
  </h3>
  <ul class="nhsuk-list nhsuk-list--tick">
    <li>
      <svg class="nhsuk-icon nhsuk-icon__tick"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path stroke-width="4"
              stroke-linecap="round"
              stroke="#007f3b"
              fill="none"
              d="M18.4 7.8l-8.5 8.4L5.6 12"
        >
        </path>
      </svg>
      cover blisters that are likely to burst with a soft plaster or dressing
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__tick"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path stroke-width="4"
              stroke-linecap="round"
              stroke="#007f3b"
              fill="none"
              d="M18.4 7.8l-8.5 8.4L5.6 12"
        >
        </path>
      </svg>
      wash your hands before touching a burst blister
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__tick"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path stroke-width="4"
              stroke-linecap="round"
              stroke="#007f3b"
              fill="none"
              d="M18.4 7.8l-8.5 8.4L5.6 12"
        >
        </path>
      </svg>
      allow the fluid in a burst blister to drain before covering it with a plaster or dressing
    </li>
  </ul>
</div>
`;

exports[`Components/DoDontList Dont smoke-test 1`] = `
<div class="nhsuk-do-dont-list">
  <h3 class="nhsuk-do-dont-list__label">
    Don't
  </h3>
  <ul class="nhsuk-list nhsuk-list--cross">
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      do not burst a blister yourself
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      do not peel the skin off a burst blister
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      do not pick at the edges of the remaining skin
    </li>
    <li>
      <svg class="nhsuk-icon nhsuk-icon__cross"
           xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           aria-hidden="true"
           height="34"
           width="34"
      >
        <path d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
        <path d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
              fill="#d5281b"
        >
        </path>
      </svg>
      do not wear the shoes or use the equipment that caused your blister until it heals
    </li>
  </ul>
</div>
`;
